x/ibc-rate-limit/ibc_middleware_test.go

```diff
@@ -0,0 +1,747 @@
+}
+
+func (suite *MiddlewareTestSuite) FullSendAToC(msg sdk.Msg) (*abci.ExecTxResult, string, error) {
+	sendResult, err := suite.SendMsgsNoCheck(suite.ChainA, msg)
+	if err != nil {
+		return nil, "", err
+	}
+
+	packet, err := ibctesting.ParsePacketFromEvents(sendResult.GetEvents())
+	if err != nil {
+		return nil, "", err
+	}
+
+	err = suite.TransferPathAC.EndpointB.UpdateClient()
+	if err != nil {
+		return nil, "", err
+	}
+
+	res, err := suite.TransferPathAC.EndpointB.RecvPacketWithResult(packet)
+	if err != nil {
+		return nil, "", err
+	}
+
+	ack, err := ibctesting.ParseAckFromEvents(res.GetEvents())
+	if err != nil {
+		return nil, "", err
+	}
+
+	err = suite.TransferPathAC.EndpointA.UpdateClient()
+	if err != nil {
+		return nil, "", err
+	}
+	err = suite.TransferPathAC.EndpointB.UpdateClient()
+	if err != nil {
+		return nil, "", err
+	}
+
+	return sendResult, string(ack), nil
+}
+
+func (suite *MiddlewareTestSuite) AssertReceive(success bool, msg sdk.Msg) (string, error) {
+	_, ack, err := suite.FullSendBToA(msg)
+	if success {
+		suite.Require().NoError(err)
+		suite.Require().NotContains(ack, "error",
+			"acknowledgment is an error")
+	} else {
+		suite.Require().Contains(ack, "error",
+			"acknowledgment is not an error")
+		suite.Require().Contains(ack, fmt.Sprintf("ABCI code: %d", types.ErrRateLimitExceeded.ABCICode()),
+			"acknowledgment error is not of the right type")
+	}
+	return ack, err
+}
+
+func (suite *MiddlewareTestSuite) AssertSend(success bool, msg sdk.Msg) (*abci.ExecTxResult, error) {
+	r, _, err := suite.FullSendAToB(msg)
+	if success {
+		suite.Require().NoError(err, "IBC send failed. Expected success. %s", err)
+	} else {
+		suite.Require().Error(err, "IBC send succeeded. Expected failure")
+		suite.ErrorContains(err, types.ErrRateLimitExceeded.Error(), "Bad error type")
+	}
+	return r, err
+}
+
+func (suite *MiddlewareTestSuite) BuildChannelQuota(name, channel, denom string, duration, sendPercentage, recvPercentage uint32) string {
+	return fmt.Sprintf(`
+          {"channel_id": "%s", "denom": "%s", "quotas": [{"name":"%s", "duration": %d, "send_recv":[%d, %d]}] }
+    `, channel, denom, name, duration, sendPercentage, recvPercentage)
+}
+
+func (suite *MiddlewareTestSuite) BuildChannelQuotaWith2Quotas(name, channel, denom string, duration1 uint32, name2 string, sendPercentage1, recvPercentage1, duration2, sendPercentage2, recvPercentage2 uint32) string {
+	return fmt.Sprintf(`
+          {"channel_id": "%s", "denom": "%s", "quotas": [{"name":"%s", "duration": %d, "send_recv":[%d, %d]},{"name":"%s", "duration": %d, "send_recv":[%d, %d]}] }
+    `, channel, denom, name, duration1, sendPercentage1, recvPercentage1, name2, duration2, sendPercentage2, recvPercentage2)
+}
+
+// Tests
+
+// Test that Sending IBC messages works when the middleware isn't configured
+func (suite *MiddlewareTestSuite) TestSendTransferNoContract() {
+	suite.ConfigureTransferChannel()
+	one := sdkmath.NewInt(1)
```

Comment:
it's `NewInt(2)` used in the [fullSendTest](https://github.com/neutron-org/neutron/blob/c4358887889516185c6ee7077a41f18c2c24f0e2/x/ibc-rate-limit/ibc_middleware_test.go#L318-L319) to avoid some kind of rounding issues. not sure what kind of issues are avoided by that, but I think we better use `NewInt(2)` instead of `NewInt(1)` just in case. here and in several other places